python/tests/avro/test_resolver.py

```diff
@@ -0,0 +1,137 @@
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+import pytest
+
+from iceberg.avro.reader import (
+    DoubleReader,
+    LongReader,
+    MapReader,
+    StringReader,
+    StructReader,
+)
+from iceberg.avro.resolver import ResolveException, resolve
+from iceberg.schema import Schema
+from iceberg.types import (
+    DoubleType,
+    IntegerType,
+    ListType,
+    LongType,
+    MapType,
+    NestedField,
+    StringType,
+    StructType,
+)
+
+
+def test_resolver():
+    write_schema = Schema(
+        NestedField(1, "id", LongType()),
+        NestedField(2, "data", StringType()),
+        NestedField(
+            3,
+            "location",
+            StructType(
+                NestedField(4, "lat", DoubleType()),
+                NestedField(5, "long", DoubleType()),
+            ),
+        ),
+        NestedField(6, "preferences", MapType(7, StringType(), 8, StringType())),
+        schema_id=1,
+    )
+    read_schema = Schema(
+        NestedField(
+            3,
+            "location",
+            StructType(
+                NestedField(4, "lat", DoubleType()),
+                NestedField(5, "long", DoubleType()),
+            ),
+        ),
+        NestedField(1, "id", LongType()),
+        NestedField(6, "preferences", MapType(7, StringType(), 8, StringType())),
+        schema_id=1,
+    )
+    read_tree = resolve(write_schema, read_schema)
+
+    assert read_tree == StructReader(
+        (
+            (1, LongReader()),
+            (None, StringReader()),
+            (
+                0,
+                StructReader(
+                    (
+                        (0, DoubleReader()),
+                        (1, DoubleReader()),
+                    )
+                ),
+            ),
+            (2, MapReader(StringReader(), StringReader())),
+        )
+    )
+
+
+def test_resolver_new_required_field():
+    write_schema = Schema(
+        NestedField(1, "id", LongType()),
+        schema_id=1,
+    )
+    read_schema = Schema(
+        NestedField(1, "id", LongType()),
+        NestedField(2, "data", StringType(), required=True),
+        schema_id=1,
+    )
+
+    with pytest.raises(ResolveException) as exc_info:
+        resolve(write_schema, read_schema)
+
+    assert "2: data: optional string is in not in the write schema, and is required" in str(exc_info.value)
+
+
+def test_resolver_invalid_evolution():
+    write_schema = Schema(
+        NestedField(1, "id", LongType()),
+        schema_id=1,
+    )
+    read_schema = Schema(
+        NestedField(1, "id", IntegerType()),
+        schema_id=1,
+    )
+
+    with pytest.raises(ResolveException) as exc_info:
+        resolve(write_schema, read_schema)
+
+    assert "Promotion from int to long is not allowed" in str(exc_info.value)
```

Comment:
It would be good to test all of the primitive type promotion cases.

Reply:
Thanks, that just uncovered a bug 👍🏻